lib/queuePopulator/IngestionReader.js

```diff
@@ -0,0 +1,233 @@
+            raftSession: raftId,
+            logger });
+        super({ zkClient, kafkaConfig, logConsumer, logId: `raft_${raftId}`,
+                logger, extensions, metricsProducer });
+        this.raftId = raftId;
+        this._iProducer = new IngestionProducer(bucketdConfig, qpConfig,
+            s3Config);
+        this.bucketdPrefix = bucketdConfig.prefix ? bucketdConfig.prefix :
+            bucketdConfig.name;
+        this.newIngestion = false;
+        this.removeLogOffset = null;
+        if (ingestionPath) {
+            this.pathToLogOffset = `${ingestionPath}` +
+                `${bucketdConfig.zookeeperSuffix}` +
+                `/logState/raft_${raftId}/logOffset`;
+        }
+    }
+
+    /* eslint-disable no-param-reassign */
+
+    _processReadRecords(params, batchState, done) {
+        const readOptions = {};
+        if (this.logOffset !== undefined) {
+            readOptions.startSeq = this.logOffset;
+        }
+        if (params && params.maxRead !== undefined) {
+            readOptions.limit = params.maxRead;
+        }
+        this.log.debug('reading records', { readOptions });
+        return async.waterfall([
+            next => this._readLogOffset((err, res) => {
+                const offset = Number.parseInt(res, 10);
+                if (err) {
+                    return next(err);
+                }
+                if (offset === 1) {
+                    this.newIngestion = true;
+                }
+                return next();
+            }),
+            next => {
+                if (!this.newIngestion) {
+                    return next();
+                }
+                return this._iProducer.snapshot(this.raftId, (err, res) => {
+                    if (err) {
+                        this.log.error('error generating snapshot for ' +
+                        'ingestion', { err });
+                        return next(err);
+                    }
+                    batchState.logRes = { info: { start: 1 }, log: res };
+                    return next();
+                });
+            },
+        ], done);
+    }
+
+    _processLogEntry(batchState, record, entry, bucketdPrefix) {
+        // for a "del", entry.value will not exist but we still need to
+        // pass through the event
+        // for a bucket metadata entry from s3Connector, there will be no
+        // entry.key but there will be an entry.type and we
+        // need to pass this through
+        if (entry.key === undefined && entry.type === undefined) {
+            return;
+        }
+        if (!record.db) {
+            this._extensions.forEach(ext => ext.filter({
+                type: entry.type,
+                bucket: entry.bucket,
+                key: entry.key,
+                value: entry.value,
+            }));
+        } else {
+            if (record.db === 'users..bucket') {
+                const keySplit = entry.key.split('..|..');
+                entry.key =
```

Comment:
There seems to be some black magic happening here that could benefit to be clarified (comments, helper functions, use of constants), though I know metadata does some black magic too. Not critical for this PR (but I give up trying to get it now :rofl:)